scenarios.md

```diff
@@ -31,6 +31,53 @@ Notes:
 | `https://e.com`<br>w/ OI   | `https://a.e.com`<br>w/o OI | `https://b.e.com`<br>w/ OI | Main in `Origin{https://e.com}`<br>A in `Site{https://e.com}`<br>B in `Origin{https://b.e.com}` |
 | `https://e.com`<br>w/o OI  | `https://a.e.com`<br>w/o OI | `https://b.e.com`<br>w/ OI | Main and A in `Site{https://e.com}`<br>B in `Origin{https://b.e.com}` |
 
+
+## Session history scenarios
+
+The previous scenarios illustrate some cases where the proposed algorithm ignores the origin isolation request because a pre-existing site-keyed agent cluster exists for a same-origin URL in the current frame tree. The following scenario illustrates a case where the session history contributes to this calculation.
+
+### Navigating a subframe
+
+* Main frame: `https://e.com` w/ OI
+* Subframe: `https://x.e.com` w/o OI
+* Outcome (per above): Main in `Origin{https://e.com}` / Sub in `Site{https://e.com}`
+
+The user clicks a link in the `https://x.e.com` subframe which takes them to `https://e.org`.
+
+The user then clicks a button in the main frame which inserts a second subframe, pointing at `https://x.e.com`. But in the meantime the server operator has deployed origin isolation on `https://x.e.com`. Thus the scenario is now:
+
+* Main frame: `https://e.com` w/ OI
+* Subframe 1: `https://e.org` (with `https://x.e.com` in the session history)
+* Subframe 2: `https://x.e.com` w/ OI
+
+In this case, subframe 2 ends up keyed by `Site{https://e.com}`. This ensures that if the user navigates subframe 1 back, restoring the `Site{https://e.com}`-keyed instance of `https://x.e.com` in subframe 1, that subframe 1 and subframe 2 are still in the same `Site{https://e.com}` agent cluster, i.e. we have avoided isolating same-origin pages from each other.
```

Comment:
Since there's no bfcache for subframes, perhaps this isn't needed? In part this logic kinda depends on a solid understanding and formalization of bfcache and BCG replacement, but I might be overlooking something.

Reply:
So there is no bfcache, but there is the HTTP cache.

Consider if a user clicked the back button in such a state, where subframe 2 was in the `Origin{https://e.com}` agent cluster. To avoid putting same-origin pages into separate agent clusters, we'd then need to put subframe 1's copy of `https://x.e.com` into the `Origin{https://e.com}` agent cluster---despite the headers for that load not requesting origin isolation.

Maybe that's OK, you might think. After all, sometimes we will origin-isolate things when their headers don't request it. (For example, in some of the simpler three-document scenarios above you can see this.) But, we do that because there was a previously-seen origin-isolated instance. Doing it in this case would be weird; we're effectively taking our cues from the second time we saw `https://x.e.com`, not the first.

So the intuition here is basically that the first time we see an origin in a BCG, that should govern things going forward.